### PR TITLE
NotificationsList: Use native sort/header functions

### PR DESCRIPTION
**Pull request by danirabbit**
todo:
 - [ ] why are headers being added multiple times

### CODE REVIEW

File: src/Widgets/NotificationsList.vala [hunk 1]
```diff
@@ -52,29 +51,58 @@ public class Notifications.NotificationsList : Gtk.ListBox {
...
+            }
+        } else {
+            row.set_header (app_entries[row_id]);
+        }
+    }
+
+    [CCode (instance_pos = -1)]
+    private int sort_func (Gtk.ListBoxRow row1, Gtk.ListBoxRow row2) {
+        var notification1 = ((NotificationEntry) row1).notification;
+        var notification2 = ((NotificationEntry) row2).notification;
```
Comment (tintou) [inline src/Widgets/NotificationsList.vala:86]:
```suggestion
        unowned NotificationEntry notification1 = ((NotificationEntry) row1).notification;
        unowned NotificationEntry notification2 = ((NotificationEntry) row2).notification;
```

File: src/Widgets/NotificationsList.vala [hunk 2]
```diff
@@ -52,29 +51,58 @@ public class Notifications.NotificationsList : Gtk.ListBox {
...
+            invalidate_sort ();
+            invalidate_headers ();
 
             Session.get_instance ().add_notification (notification);
         }
     }
 
+    [CCode (instance_pos = -1)]
+    private void header_func (Gtk.ListBoxRow row, Gtk.ListBoxRow? before) {
+        var row_id = ((NotificationEntry) row).notification.desktop_id;
```
Comment (tintou) [inline src/Widgets/NotificationsList.vala:70]:
```suggestion
        unowned string row_id = ((NotificationEntry) row).notification.desktop_id;
```

File: src/Widgets/NotificationsList.vala [hunk 3]
```diff
@@ -52,29 +51,58 @@ public class Notifications.NotificationsList : Gtk.ListBox {
...
+        if (before != null) {
+            var before_id = ((NotificationEntry) before).notification.desktop_id;
```
Comment (tintou) [inline src/Widgets/NotificationsList.vala:72]:
```suggestion
            unowned string before_id = ((NotificationEntry) before).notification.desktop_id;
```

File: src/Widgets/NotificationsList.vala [hunk 4]
```diff
@@ -37,7 +34,9 @@ public class Notifications.NotificationsList : Gtk.ListBox {
 
         activate_on_single_click = true;
         selection_mode = Gtk.SelectionMode.NONE;
+        // set_header_func (header_func);
```
Comment (tintou) [inline src/Widgets/NotificationsList.vala:37]: Ideally you can put this after the `set_sort_func`

File: src/Widgets/NotificationsList.vala [hunk 5]
```diff
@@ -52,29 +51,58 @@ public class Notifications.NotificationsList : Gtk.ListBox {
...
-                int insert_pos = table.get (app_entry.app_id);
-                insert (entry, insert_pos + 1);
             }
+            add (entry);
 
             app_entry.clear.connect (clear_app_entry);
 
             show_all ();
+            invalidate_sort ();
+            invalidate_headers ();
```
Comment (tintou) [inline src/Widgets/NotificationsList.vala:62]: `invalidate_sort` already calls `invalidate_headers` ( https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/gtk/gtklistbox.c#L1328 )

**Review by tintou — CHANGES_REQUESTED** [review]

File: src/Widgets/NotificationsList.vala (hunk 5 above)
Comment (danrabbit) [reply src/Widgets/NotificationsList.vala:62]: wow TIL

**Pull request closed without merging**